### PR TITLE
[babel 8] fix properties name for function-like TS nodes

### LINKED ISSUE #9231: Typescript: Inconsistent properties in TS function like nodes [parser, estree]
## Bug Report

**Current Behavior**
Ast structure is inconsistent for function like nodes. 

### requested change
- property `params` should be renamed to `parameters`
- property `typeAnnotation` should be renamed to `returnType` if used as return type

### affected nodes
 - `TSCallSignatureDeclaration`
 - `TSConstructSignatureDeclaration`
 - `TSFunctionType`
 - `TSConstructorType`
 - `TSMethodSignature`

**Input Code**
```ts
interface A {
  (a: string): string
  new(a: string): string
}
```

Other function like nodes (FunctionExpression, ArrowFunctionExpression, FunctionDeclaration, ....) have properties:
- `typeParameters`
- `params`
- `returnType`

**Estree**

https://github.com/estree/estree/blob/master/extensions/type-annotations.md#functions

**Environment**
- Babel version(s): 7.2.3
- Node/npm version: 6-11
- OS: win/linux
- Monorepo: no
- How you are using Babel: cli

### PR DESCRIPTION
**Pull request by sosukesuzuki**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #9231
| Major: Breaking Change?  | Y
| Tests Added + Pass?      | Yes
| License                  | MIT

**This PR contains many test changes. It is recommended that you review each commit.**

Change `parameters` and `typeAnnotation` to `params` and `returnType` in the following 5 function-like nodes

- `TSCallSignatureDeclaration`
- `TSMethodSignature`
- `TSConstructSignatureDeclaration`
- `TSFunctionType`
- `TSConstructorType`

**Code examples:**

```ts
interface Foo {
  // TSCallSignatureDeclaration
  <T>(): string;

  // TSMethodSignature
  foo<T>(): string;

  // TSConstructSignatureDeclaration
  new <T>(): string;
}

// TSFunctionType
type Bar = <T>() => string;

// TSConstructorType
type Baz = new <T>() => string;
```

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13709"></a>

### CODE REVIEW

File: packages/babel-generator/src/generators/typescript.ts [hunk 1]
```diff
@@ -221,15 +221,24 @@ export function tsPrintFunctionOrConstructorType(
...
   this.space();
   this.token("=>");
   this.space();
-  this.print(node.typeAnnotation.typeAnnotation, node);
+  const returnType = process.env.BABEL_8_BREAKING
+    ? node.returnType
+    : node.typeAnnotation;
+  if (!returnType) {
+    console.log(node);
+  }
```
Comment (fedeci) [inline packages/babel-generator/src/generators/typescript.ts:240]:
```suggestion
```

File: packages/babel-types/src/definitions/typescript.ts [hunk 2]
```diff
@@ -91,13 +91,20 @@ defineType("TSQualifiedName", {
 
 const signatureDeclarationCommon = {
   typeParameters: validateOptionalType("TSTypeParameterDeclaration"),
-  parameters: validateArrayOfType(["Identifier", "RestElement"]),
-  typeAnnotation: validateOptionalType("TSTypeAnnotation"),
+  [process.env.BABEL_8_BREAKING ? "params" : "parameters"]: validateArrayOfType(
```
Comment (fedeci) [inline packages/babel-types/src/definitions/typescript.ts:94]: Q: don't we use BABEL_TYPES_8_BREAKING in babel-types?

**Review by fedeci — APPROVED** [review]

File: packages/babel-types/src/definitions/typescript.ts (hunk 2 above)
Comment (sosukesuzuki) [reply packages/babel-types/src/definitions/typescript.ts:94]: Oh I didn't know `BABEL_TYPES_8_BREAKING`  environment variables.

Comment (nicolo-ribaudo) [reply packages/babel-types/src/definitions/typescript.ts:94]: I prefer `BABEL_8_BREAKING` here. `BABEL_TYPES_8_BREAKING` can be used by `@babel/types` consumers to opt-in into stricter validation that we will have in Babel 8. However, opting-in into this AST change makes it impossible to use `@babel/types` with `@babel/parser` and `@babel/generator`.

File: packages/babel-types/test/fields.js [hunk 3]
```diff
@@ -52,6 +52,26 @@ const ignoredFields = {
   ObjectMethod: { method: true, id: true, predicate: true },
   StaticBlock: { static: true },
   TSDeclareMethod: { id: true },
+  // TODO(Babel 8): Remove
+  // https://github.com/babel/babel/issues/9231
+  ...(process.env.BABEL_8_BREAKING
```
Comment (nicolo-ribaudo) [inline packages/babel-types/test/fields.js:57]:
```suggestion
  ...(process.env.BABEL_8_BREAKING
```

No need for the comment, it's implicit in the `process.env.BABEL_8_BREAKING` check.

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by sosukesuzuki** [comment]
@JLHwung Could you review this?

**Review by JLHwung — APPROVED** [review]

**Pull request merged**